# Summary

* [Preface](README.md)
* [Getting Started](getting-started.md)

## Part I. Basic Tools

* [1.Linux](part-i.-basic-tools/1.linux.md)
* [2.Blast](part-i.-basic-tools/2.seqblast.md)
* [3.Conservation Analysis](part-i.-basic-tools/3.conservation.md)
* [4. Function Analysis](part-i.-basic-tools/4.gokegg/README.md)
  * [4.1 GO](part-i.-basic-tools/4.gokegg/4.1.go.md)
  * [4.2 KEGG](part-i.-basic-tools/4.gokegg/4.2.kegg.md)
  * [4.3 GSEA](part-i.-basic-tools/4.gokegg/43-gsea.md)

## Part II. NGS Pipelines

* [5.Mapping](part-ii.-ngs-pipelines/5.seqbowtie.md)
* [6.RNA-seq](part-ii.-ngs-pipelines/6.differential_alternative/README.md)
  * [6.1 Differential Expression](part-ii.-ngs-pipelines/6.differential_alternative/6.1.basic_analyses.md)
  * [6.2 Alternative Splicing](part-ii.-ngs-pipelines/6.differential_alternative/6.2.alternative.md)
* [7.ChIP-seq](part-ii.-ngs-pipelines/7.chipseq_motif.md)
  * [7.1 Peak Calling](part-ii.-ngs-pipelines/71-peak-calling.md)
  * [7.2 Motif Analysis](part-ii.-ngs-pipelines/72-motif-analysis.md)

## Appendix

* [Appendix I. Keep Learning](appendix/appendix-more.md)
* [Appendix II. Databases & Servers](appendix/appendix-databases.md)
* [Appendix III. Scripts & Plots](appendix/appendix-shared-scripts.md)

