# 2. Blast

## Pipeline
![](/.gitbook/assets/blast-pipeline.png)

## Data Structure


  了解存储sequence的常用文件格式**

**FASTA格式** （_.fasta or_ .fa )

&gt;gi\|47115317\|emb\|CAG28618.1\| VIM \[Homo sapiens\] MSTRSVSSSSYRRMFGGPGTASRPSSSRSYVTTSTRTYSLGSALRPSTSRSLYASSPGGVYATRSSAVRL

`The word following the ">" symbol is the identifier of the sequence, and the rest of the line is the description (optional). Normally, identifiers are simply protein accession, name or Entrez gi's (e.g., Q5I7T1, AG10B_HUMAN, 129295), but a bar-separated NCBI sequence identifier (e.g., gi|129295) will also be accepted. Any arbitrary user-specified sequence identifier can also be used (e.g., CLONE00073452).`

query protein sequence 如下 (Homo sapiens CAG28618.1)

```
>gi|47115317|emb|CAG28618.1| VIM [Homo sapiens]
MSTRSVSSSSYRRMFGGPGTASRPSSSRSYVTTSTRTYSLGSALRPSTSRSLYASSPGGVYATRSSAVRL
```

#### Inputs

| Format | Description                     | Notes                    |
|--------|---------------------------------|--------------------------|
| fasta  | contains gene name and sequence | gene name start with `>` |

#### Outputs

| Format | Description               | Notes |
|--------|---------------------------|-------|
| blastp | query aligned to database |       |

## Blast in local

**1. 安装blast及其子程序blastn, blastp，blastx, ...** 

> **Note: Docker 中已经装好 **

_安装方式1-自动安装:_

Linux自动安装软件方法： sudo apt-get install 软件名称

```bash
sudo apt-get install ncbi-blast+        #密码：cs
```

_安装方式2-手动安装:_

[Download link](ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/)

寻找类似如下文件：

a. 32位计算机（老机器）安装文件：如 ncbi-blast-2.2.28+-ia32-linux.tar.gz

b. 64位机器安装文件： 如 ncbi-blast-2.2.28+-x64-linux.tar.gz

（可以通过 uname –a 查看机器类型是64还是32位）




>```bash 
blastn –h #阅读命令帮助
blastn --help
```


以下步骤均在 `/home/test/rna-seq/` 下进行:  
`cd /home/test/rna-seq/`

**II. 两条序列比对**

_**Protein sequence alignment**_


利用blastp进行蛋白质比对

```bash
blastp  -query protein/VIM.fasta  -subject   protein/NMD.fasta   -out output/blastp
```

VIM.fasta 与 NMD.fasta 分别是金属beta酶家族的两个亚种酶的序列

_**DNA sequence alignment**_

利用blastn进行DNA序列比对

当前工作目录：/home/cs/Bioinfo\_Lab/1.Sequence/1.seq\_align/blast/dna

```bash
blastn  -query dna/H1N1-HA.fasta  -subject   dna/H7N9-HA.fasta   -out output/blastn
```

H1N1-HA.fasta 与H7N9-HA.fasta 是流感病毒序列文件

**III.  一条序列与远程序列库比对**

 ****_**Protein sequence alignment 比对远程pdb数据库**_

当前工作目录：/home/cs/Bioinfo\_Lab/1.Sequence/1.seq\_align/blast/protein/

```bash
blastp  -query protein/VIM.fasta  -db pdb -remote   -out output/blastp_remote
```

 _**DNA sequence alignment 比对远程nr数据库**_

当前工作目录：/home/cs/Bioinfo\_Lab/1.Sequence/1.seq\_align/blast/dna

```
blastn  -query dna/H1N1-HA.fasta  -db nr  -remote   -out output/blastn_remote
```

**IV. 一条序列在自定义的序列库里比对**

例如：在yeast基因组序列中搜索Yeast.fasta序列

当前工作目录：/home/cs/Bioinfo\_Lab/1.Sequence/1.seq\_align/blast／

**Step 1: 建库**

```bash
makeblastdb -dbtype nucl -in dna/YeastGenome.fa -out database/YeastGenome
```

- `-dbtype`: 待建库的类型（`nucl`, `prot`)
- `-in`: 待建库的序列文件
- `-out`: 序列库名前缀

**Step 2: 比对**

```bash
blastn -query dna/Yeast.fasta -db database/YeastGenome -out output/Yeast.blastn
```

##  Blast in web


![](../.gitbook/assets/blastweb.png)





### Running Steps

#### **Input sequence**

![](../.gitbook/assets/blastweb2.png)

#### **Select database**

![](../.gitbook/assets/blastweb3.png)

#### **Select algorithm**

![](../.gitbook/assets/blastweb4.png)

#### **Output results**

![](../.gitbook/assets/blastweb5.png)

#### **Output table**

![](../.gitbook/assets/blastweb6.png)

#### **Alignment details**

![](../.gitbook/assets/blastweb7.png)

### Tips/Utilities

**A better view of fasta file**

```bash
less -S gi47115317.fa  # chop long lines rather than wrap them
```

### Homework and more

**1 Blast sequence with mouse genome.**

Blast protein sequence only with mouse genome protein database, and only keep the max target sequences is 10, expect threshold is 0.5.

