# 6.1 Differential expression

**Next-Generation Sequencing \(NGS\) data: RNA-seq**

## I. 上机指南

### 0. preparation

_**0.1 Data Preparation**_ _\*\*_

The Yeast RNA-seq data were downloaded from GSE42983, random sample 100K or 10K reads per sample,

_• wild-type samples: two replicates \(wt1.fq and wt2.fq\)_

_• H2O2 treatment \(oxidative stress\) samples: two replicates \(wt1X.fq and wt2X.fq\)_

文件位置：/home/cs/Bioinfo\_Lab/3.Genomics/Diff\_Express/Raw\_reads\_10k

You can also download the files by click [here](https://www.jianguoyun.com/p/DTCJc-gQ0NLuBRj9kQ4#dir=%2F2016%2F生物信息学导论%2F上机文件%2F3.Genomics::mode=0).

_**0.2 Environment Preparation**_

安装R软件包

网络学堂或点击[链接](https://www.jianguoyun.com/p/DTCJc-gQ0NLuBRj9kQ4#dir=%2F2016%2F生物信息学导论%2F上机文件%2F3.Genomics::mode=0)，下载RPackages，并放到/home/cs/Bioinfo\_Lab/3.Genomics/bin/目录下，然后:

```text
cd /home/cs/Bioinfo_Lab/3.Genomics/bin/

tar –zxvf RPackages.tar.gz         

source bin_path_R

cd ..

cd Diff_Express
```

### 1. Align the RNA-seq reads to the genome

**1.1 map reads using Tophat**

TopHat maps short sequences from spliced transcripts to whole genomes.

语法：`tophat [options] <bowtie_index> <reads1[,reads2,...]> [reads1[,reads2,...]] \[quals1,[quals2,...]] [quals1[,quals2,...]］`

命令：

```text
tophat -p 4 -G yeast_annotation.gff --no-coverage-search -o wt1_thout bowtie_index/YeastGenome Raw_reads_10k/wt1.fq 

tophat -p 4 -G yeast_annotation.gff --no-coverage-search -o wt2_thout bowtie_index/YeastGenome Raw_reads_10k/wt2.fq 

tophat -p 4 -G yeast_annotation.gff --no-coverage-search -o wt1X_thout bowtie_index/YeastGenome Raw_reads_10k/wt1X.fq 

tophat -p 4 -G yeast_annotation.gff --no-coverage-search -o wt2X_thout  bowtie_index/YeastGenome Raw_reads_10k/wt2X.fq
```

_-p/--num-threads  \[ default: 1 \]_

_-G/--GTF  \(GTF/GFF with known transcripts\)_

_-o/--output-dir  \[ default: ./tophat\_out \]_

**1.2 Extract mapped reads on chr I only \(for quick running\)**

**index bam file**

```text
bamtools index -in wt1_thout/accepted_hits.bam 

bamtools index -in wt2_thout/accepted_hits.bam 

bamtools index -in wt1X_thout/accepted_hits.bam

bamtools index -in wt2X_thout/accepted_hits.bam
```

_index:Generates index for BAM file_

**extract**

```text
bamtools filter -region chrI -in wt1_thout/accepted_hits.bam -out wt1_thout/chrI.bam

bamtools filter -region chrI -in wt2_thout/accepted_hits.bam -out wt2_thout/chrI.bam

bamtools filter -region chrI -in wt1X_thout/accepted_hits.bam -out wt1X_thout/chrI.bam

bamtools filter -region chrI -in wt2X_thout/accepted_hits.bam -out wt2X_thout/chrI.bam
```

_filter: Filters BAM files by user-specified criteria_

### 2. Assemble transcripts on Chr I by Cufflinks

**2.1 Assemble transcripts for each sample:**

```text
cufflinks -p 4 -o wt1_clout wt1_thout/chrI.bam 

cufflinks -p 4 -o wt2_clout wt2_thout/chrI.bam 

cufflinks -p 4 -o wt1X_clout wt1X_thout/chrI.bam 

cufflinks -p 4 -o wt2X_clout wt2X_thout/chrI.bam
```

**2.2 Create a file called assemblies.txt that lists the assembly files for each sample.**

The file should contain the following lines:

_wt1X\_clout/transcripts.gtf_

_wt1\_clout/transcripts.gtf_

_wt2X\_clout/transcripts.gtf_

_wt2\_clout/transcripts.gtf_

You can use vi to create a file containing this command:

```text
rm -f assemblies.txt;
for i in `ls | grep clout`;
do 
  echo ./$i/transcripts.gtf >> assemblies.txt;
done
```

**2.3 Merge all assemblies to one file containing merged transcripts:**

cuffmerge takes two or more Cufflinks GTF files and merges them into a single unified transcript catalog

`cuffmerge -g yeast_chrI_annotation.gff -s bowtie_index/YeastGenome.fa -p 4 assemblies.txt`

### 3. Identify differentially expressed genes and transcripts

\(we’ll show the output from 1M\_reads, not 10K\_reads\)

Run cuffdiff based on the merged transcripts \(assemblies.txt\) and mapped reads for each sample \(\*.bam\)

`cuffdiff -o diff_out -b bowtie_index/YeastGenome.fa -p 4 -u merged_asm/merged.gtf ./wt1_thout/chrI.bam,./wt2_thout/chrI.bam ./wt1X_thout/chrI.bam,./wt2X_thout/chrI.bam`

This won’t generate sufficient result from 10K reads, see diff\_out of 1M reads in examples/plot\_example\_1Mreads/diff\_out diff\_out

### 4. Explore differential analysis results with CummeRbund TIMING variable

\(we’ll use the output from 1M\_reads, not 10k reads\)

**requirements**

_Cufflinks &gt;= v2.0.0_

_SQLite_

_R &gt;= v2.7.0_

**Required Packages:**

_RSQLite_

_ggplot2 v0.9.2_

_reshape2_

_plyr_

_fastcluster_

_rtracklayer_

_Gviz_

_BiocGenerics \(&gt;=0.3.2\)_

**Recommended Package:**

_Hmisc_

命令如下：

```text
Rscript  plot_DE_chart.R       #See example figures of 1M reads in examples/plot_example_1Mreads/DE_plots
```

## II.上机任务

1.理解和掌握基本的基因差异表达分析 方法（Differential Expression Analysis）; 学会Differential Expression Analysis的基本软件。

2.按照Instruction上的方法，使用 TopHat 和Cufflinks完成Differential Expression Analysis。

## III.上机作业

1. 阅读： Trapnell C et al. Nat Protoc. 2012 Mar 1;7\(3\):562-78 Differential gene and transcript expression analysis of RNA-seq experiments with TopHat and Cufflinks. [Reference Link](http://www.ncbi.nlm.nih.gov/pubmed/22383036)
2. 画一张Plot展示differentially expressed transcripts, 并写一段话解释Plot
3. 寻找differentially expressed transcripts，要对数据做怎样的处理？这些处理有哪些统计学上的考虑？



