# 6.1 Differential expression



本章我们将

1. 理解和掌握基本的基因差异表达分析方法（Differential Expression Analysis）；学会 Differential Expression Analysis 的基本软件。

2. 使用 **TopHat** 和 **Cufflinks** 完成 Differential Expression Analysis。


## 1) Pipeline


## 2) Data Structure

### 2a) getting software & data (already available in Docker)

(1) software

    1. R & **cummeRbund** package (BioConductor) 
    1. `tophat`
    1. `bamtools`
    1. `cufflinks`

(2) Data

   The Yeast RNA-seq data were downloaded from [GSE42983](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE42983), random sample 1M or 10K reads per sample,
   
   - wild-type samples: two replicates, `wt1.fq` and `wt2.fq`
   
   - H2O2 treatment _oxidative stress_ samples: two replicates, `wt1X.fq` and `wt2X.fq`
   
   文件位置：`rna-seq/Raw_reads_10k` (You can also download them [here](https://github.com/dongzhuoer/lulab-teaching_book-data/blob/master/DE/data.tar.gz) )
   

### 2b) input

| **format** | **description** | **Notes** |
| :--- | :--- | :--- | 
| .fastq |RNA sequences of each sample| raw data of RNA-seq | 
|.fa | sequences of the whole genome  |  | 
| .gff| genome annotation file | Default genome annotation file is from GENCODE, is similar to gtf file |
|  | bowtie index file | used to align RNA sequences to genome|

### 2c) output

| **format** | **description** | **Notes** |
| :--- | :--- | :--- | 
|.bam | genome mapped reads (binary format of sam)|generated by align step  |
|transcript.gtf|assembled transcripts of each sample|generated by assemble step|
|merged.gtf| all annotation of assembled transcripts|generated by merge step|
||differentially expressed genes|generated by DE gene identify step|
||R plot to explore differentially expressed genes |generated by R package|

## 3) Running Steps

以下步骤均在 `/home/test/rna-seq/` 下进行:  
`cd /home/test/rna-seq/`

### 3a) Align the RNA-seq reads to the genome

(1) map reads using Tophat

   `tophat` maps short sequences from spliced transcripts to whole genome.

   ```bash
   tophat -p 4 -G yeast_annotation.gff --no-coverage-search -o mapping/wt1_thout  bowtie_index/YeastGenome Raw_reads_10k/wt1.fq 
   
   tophat -p 4 -G yeast_annotation.gff --no-coverage-search -o mapping/wt2_thout  bowtie_index/YeastGenome Raw_reads_10k/wt2.fq 
   
   tophat -p 4 -G yeast_annotation.gff --no-coverage-search -o mapping/wt1X_thout bowtie_index/YeastGenome Raw_reads_10k/wt1X.fq 
   
   tophat -p 4 -G yeast_annotation.gff --no-coverage-search -o mapping/wt2X_thout bowtie_index/YeastGenome Raw_reads_10k/wt2X.fq 
   ```
   
   用法：`tophat [options] <bowtie_index> <reads1[,reads2,...]> [reads1[,reads2,...]]`
      
   
   - `-p/--num-threads`  default: 1                     
   - `-G/--GTF`          GTF/GFF with known transcripts 
   - `-o/--output-dir`   default: `./tophat_out`          
   
   
   
(2) Extract mapped reads on chr I only (for quick running)
   
   - index `.bam` file
    
    `index` command generates index for `.bam` file
   
     ```bash     
     bamtools index -in mapping/wt1_thout/accepted_hits.bam 
     
     bamtools index -in mapping/wt2_thout/accepted_hits.bam 
     
     bamtools index -in mapping/wt1X_thout/accepted_hits.bam 
     
     bamtools index -in mapping/wt2X_thout/accepted_hits.bam
     ``` 
     
   - extract

    `filter` command filters `.bam` files by user-specified criteria
     
     ```bash
     bamtools filter -region chrI -in mapping/wt1_thout/accepted_hits.bam -out mapping/wt1_thout/chrI.bam
     
     bamtools filter -region chrI -in mapping/wt1_thout/accepted_hits.bam -out mapping/wt2_thout/chrI.bam
     
     bamtools filter -region chrI -in mapping/wt1_thout/accepted_hits.bam -out mapping/wt1X_thout/chrI.bam
     
     bamtools filter -region chrI -in mapping/wt1_thout/accepted_hits.bam -out mapping/wt2X_thout/chrI.bam
     ```
     

### 3b) Assemble transcripts on Chr I by Cufflinks

(1) Assemble transcripts for each sample:
   
   ```bash
   cufflinks -p 4 -o assembly/wt1_clout  mapping/wt1_thout/chrI.bam 
   
   cufflinks -p 4 -o assembly/wt2_clout  mapping/wt2_thout/chrI.bam 
   
   cufflinks -p 4 -o assembly/wt1X_clout mapping/wt1X_thout/chrI.bam 
   
   cufflinks -p 4 -o assembly/wt2X_clout mapping/wt2X_thout/chrI.bam
   ```
   
(2) Create a file called `assemblies.txt` which lists the assembly files for each sample in the `assembly` folder. Its content is as follows:
   
   ```
   assembly/_wt1X_clout/transcripts.gtf
   assembly/_wt1_clout/transcripts.gtf
   assembly/_wt2X_clout/transcripts.gtf   
   assembly/_wt2_clout/transcripts.gtf
   ```

   You can create that file manually by `vim` or use this command:  
   `ls assembly/*/transcripts.gtf > assembly/assemblies.txt`

   
 
 (3) Merge all assemblies to one file containing merged transcripts:
   
   `cuffmerge` takes two or more Cufflinks `.gtf` files and merges them into a single unified transcript catalog
   
   ```bash
   cuffmerge -g yeast_chrI_annotation.gff -s bowtie_index/YeastGenome.fa -p 4 -o assembly/merged assembly/assemblies.txt
   ```

### 3c) Identify differentially expressed genes and transcripts

we’ll use the output from 1M reads, not 10K reads

Run `cuffdiff` based on the merged transcripts and mapped reads for each sample

```bash
cuffdiff -o diff_expr -b bowtie_index/YeastGenome.fa -p 4 -u assembly/merged/merged.gtf \
    mapping/wt1_thout/chrI.bam,mapping/wt2_thout/chrI.bam mapping/wt1X_thout/chrI.bam,mapping/wt2X_thout/chrI.bam
```

This won’t generate sufficient result from 10K reads, see results of 1M reads in `1M_reads_diff_expr` 

### 3d) Explore differential analysis results with CummeRbund package

we’ll use the output from 1M reads, not 10k reads <!-- do we need SQLite (RSQLite package)? -->

```bash
Rscript  plot_DE_chart.R       # See example figures of 1M reads in 1M_reads_diff_expr
```

## 4) Tips/Utilities

More reading:  [Transcript-level expression analysis of RNA-seq experiments with HISAT, StringTie and Ballgown](https://www.nature.com/articles/nprot.2016.095)

## 5) Homework and more

(1) 阅读 Trapnell C et al. Nat Protoc. 2012 Mar 1;7(3):562-78 Differential gene and transcript expression analysis of RNA-seq experiments with TopHat and Cufflinks. [Reference Link](http://www.ncbi.nlm.nih.gov/pubmed/22383036)
(2) 画一张Plot展示 differentially expressed transcripts, 并写一段话解释Plot
(3) 寻找 differentially expressed transcripts，要对数据做怎样的处理？这些处理有哪些统计学上的考虑？



