# 6.1 Differential expression

~~**Next-Generation Sequencing \(NGS\) data: RNA-seq**~~

本章我们将

1. 理解和掌握基本的基因差异表达分析方法（Differential Expression Analysis）；学会 Differential Expression Analysis 的基本软件。

2. 使用 **TopHat** 和 **Cufflinks** 完成 Differential Expression Analysis。


## Pipeline


## Data Structure

### getting software & data

1. Data

   The Yeast RNA-seq data were downloaded from [GSE42983](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE42983), 
       random sample 100K or 10K reads per sample,
   
   - wild-type samples: two replicates, `wt1.fq` and `wt2.fq`
   
   - H2O2 treatment _oxidative stress_ samples: two replicates, `wt1X.fq` and `wt2X.fq`
   
   文件位置：`/home/cs/Bioinfo_Lab/3.Genomics/Diff_Express/Raw_reads_10k`
   
   You can also download the files by click [here](https://www.jianguoyun.com/p/DTCJc-gQ0NLuBRj9kQ4#) (`2016/生物信息学导论/上机文件/3.Genomics/3.Diff_Express.zip`).

2. Environment Preparation

   安装R软件包
   
   网络学堂或点击[链接](https://www.jianguoyun.com/p/DTCJc-gQ0NLuBRj9kQ4#) (`2016/生物信息学导论/上机文件/3.Genomics/R-packages.tar.gz`)，并放到 `/home/cs/Bioinfo_Lab/3.Genomics/bin/` 目录下，然后:
   
   ```bash
   cd /home/cs/Bioinfo_Lab/3.Genomics/bin/
   
   tar –zxvf RPackages.tar.gz         
   
   source bin_path_R
   
   cd ..
   
   cd Diff_Express
   ```

### input

### output

## Running Scripts

### Align the RNA-seq reads to the genome

1. map reads using Tophat

   TopHat maps short sequences from spliced transcripts to whole genomes.
   
   语法：`tophat [options] <bowtie_index> <reads1[,reads2,...]> [reads1[,reads2,...]] \[quals1,[quals2,...]] [quals1[,quals2,...]］`
   
   命令：
   
   ```bash
   tophat -p 4 -G yeast_annotation.gff --no-coverage-search -o wt1_thout bowtie_index/YeastGenome Raw_reads_10k/wt1.fq 
   
   tophat -p 4 -G yeast_annotation.gff --no-coverage-search -o wt2_thout bowtie_index/YeastGenome Raw_reads_10k/wt2.fq 
   
   tophat -p 4 -G yeast_annotation.gff --no-coverage-search -o wt1X_thout bowtie_index/YeastGenome Raw_reads_10k/wt1X.fq 
   
   tophat -p 4 -G yeast_annotation.gff --no-coverage-search -o wt2X_thout  bowtie_index/YeastGenome Raw_reads_10k/wt2X.fq
   ```
   
   |                    |                                |
   |--------------------|--------------------------------|
   | `-p/--num-threads` | default: 1                     |
   | `-G/--GTF`         | GTF/GFF with known transcripts |
   | `-o/--output-dir`  | default: ./tophat_out          |

     
   
2. Extract mapped reads on chr I only (for quick running)
   
   - index bam file
   
     ```bash
     bamtools index -in wt1_thout/accepted_hits.bam 
     
     bamtools index -in wt2_thout/accepted_hits.bam 
     
     bamtools index -in wt1X_thout/accepted_hits.bam
     
     bamtools index -in wt2X_thout/accepted_hits.bam
     ```
     
     `index`: Generates index for BAM file
     
   - extract
     
     ```bash
     bamtools filter -region chrI -in wt1_thout/accepted_hits.bam -out wt1_thout/chrI.bam
     
     bamtools filter -region chrI -in wt2_thout/accepted_hits.bam -out wt2_thout/chrI.bam
     
     bamtools filter -region chrI -in wt1X_thout/accepted_hits.bam -out wt1X_thout/chrI.bam
     
     bamtools filter -region chrI -in wt2X_thout/accepted_hits.bam -out wt2X_thout/chrI.bam
     ```
     
     `filter`: Filters BAM files by user-specified criteria

### Assemble transcripts on Chr I by Cufflinks

1. Assemble transcripts for each sample:
   
   ```bash
   cufflinks -p 4 -o wt1_clout wt1_thout/chrI.bam 
   
   cufflinks -p 4 -o wt2_clout wt2_thout/chrI.bam 
   
   cufflinks -p 4 -o wt1X_clout wt1X_thout/chrI.bam 
   
   cufflinks -p 4 -o wt2X_clout wt2X_thout/chrI.bam
   ```
   
2. Create a file called `assemblies.txt` that lists the assembly files for each sample:
   
   ```
   _wt1X_clout/transcripts.gtf
   _wt1_clout/transcripts.gtf
   _wt2X_clout/transcripts.gtf   
   _wt2_clout/transcripts.gtf
   ```

   You can create that file manually or use `vi`:
   
   ```bash
   rm -f assemblies.txt;
   for i in `ls | grep clout`;
   do 
     echo ./$i/transcripts.gtf >> assemblies.txt;
   done
   ```
   
3. Merge all assemblies to one file containing merged transcripts:
   
   cuffmerge takes two or more Cufflinks GTF files and merges them into a single unified transcript catalog
   
   ```bash
   cuffmerge -g yeast_chrI_annotation.gff -s bowtie_index/YeastGenome.fa -p 4 assemblies.txt
   ```

### Identify differentially expressed genes and transcripts

we’ll show the output from 1M_reads, not 10K_reads

Run cuffdiff based on the merged transcripts `assemblies.txt` and mapped reads for each sample `.bam`

```bash
cuffdiff -o diff_out -b bowtie_index/YeastGenome.fa -p 4 -u merged_asm/merged.gtf \
    ./wt1_thout/chrI.bam,./wt2_thout/chrI.bam ./wt1X_thout/chrI.bam,./wt2X_thout/chrI.bam
```

This won’t generate sufficient result from 10K reads, see `diff_out` of 1M reads in `examples/plot_example_1Mreads/diff_out` 

### Explore differential analysis results with CummeRbund TIMING variable

we’ll use the output from 1M_reads, not 10k reads

- requirements

  Cufflinks (>= v2.0.0)  
  SQLite  
  R ()>= v2.7.0)

- Required Packages:

  RSQLite  
  ggplot2 v0.9.2  
  reshape2  
  plyr  
  fastcluster  
  rtracklayer  
  Gviz  
  BiocGenerics (>=0.3.2)  

- Recommended Package:

  Hmisc

命令如下：

```bash
Rscript  plot_DE_chart.R       #See example figures of 1M reads in examples/plot_example_1Mreads/DE_plots
```

## Tips/Utilities

## Homework and more

1. 阅读 Trapnell C et al. Nat Protoc. 2012 Mar 1;7\(3\):562-78 Differential gene and transcript expression analysis of RNA-seq experiments with TopHat and Cufflinks. [Reference Link](http://www.ncbi.nlm.nih.gov/pubmed/22383036)
2. 画一张Plot展示differentially expressed transcripts, 并写一段话解释Plot
3. 寻找differentially expressed transcripts，要对数据做怎样的处理？这些处理有哪些统计学上的考虑？



