# 6.2 Alternative splicing

http://rnaseq-mats.sourceforge.net/rmats4.0.2/user_guide.htm

## 1) Pipeline

![](../.gitbook/assets/AS-pipeline.png)

## 2) Data Structure

### 2a) getting software & data

(1) 安装 [MATS](http://rnaseq-mats.sourceforge.net/rmats4.0.2/index.html) (Docker 中已装好)
(2) 我们使用 [PRJNA130865](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA130865) 中的两个样本: 
   - [SRR065545](https://www.ebi.ac.uk/ena/data/view/SRR065545): C2C12 with control shRNA vector
   - [SRR065544](https://www.ebi.ac.uk/ena/data/view/SRR065544): C2C12 with shRNA against CUGBP1  
   
   我们已经准备好 `.bam` 文件（仅包含 mapping 到 X 染色体上的部分），读者也可以点击相应链接下载原始的 FASTQ 文件） 
   
(3) 从 [Ensembl](ftp://ftp.ensembl.org/pub/release-93/gtf/mus_musculus/Mus_musculus.GRCm38.93.gtf.gz) 下载  _Mus musculus_ 的基因组注释（GRCm38/mm10）


### 2b) input

| Format | Description          | Notes |
|--------|----------------------|-------|
| `.bam` | 将样本中的 Reads 比对到参考基因组 | -     |
| `.gtf` | 参考基因组注释文件            | -     |


### output

| Format   | Description                                                            | Notes |
|----------|------------------------------------------------------------------------|-------|
| many TSV | all possible alternative splicing (AS) events derived from GTF and RNA |   -    |

详细说明请参见 <http://rnaseq-mats.sourceforge.net/rmats4.0.2/user_guide.htm#output>

## 3) Running Steps

以下步骤均在 `/home/test/AS/` 下进行:  
`cd /home/test/AS/`


```bash
echo "input/SRR065544_chrX.bam" > input/b1.txt
echo "input/SRR065545_chrX.bam" > input/b2.txt

python2 /usr/local/rMATS-turbo-Linux-UCS4/rmats.py \
    --b1 input/b1.txt --b2 input/b2.txt --gtf input/Mus_musculus_chrX.gtf 
    --od output --tstat 4 -t paired --readLength 35 \
    --cstat 0.0001 --libType fr-unstranded
```

输出文件位于 `mouse_ChrX/` 中，参数意义请参见 <http://rnaseq-mats.sourceforge.net/rmats4.0.2/user_guide.htm#using>

## 4) Tips/Utilities

### 4a) read more
读者可参考以下文献，探索其他的 alternative splicing 分析软件

- [A survey of software for genome-wide discovery of differential splicing in RNA-Seq data](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3903050/)
- [A survey of computational methods in transcriptome-wide alternative splicing analysis](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5203768/)

或者使用 [SAJR](http://storage.bioinf.fbb.msu.ru/~mazin/) in [Nature Protocol](https://www.nature.com/protocolexchange/protocols/6093/)

## 4b) generate `.bam` file

1. install **hisat**, **bamtools**, **samtools**
1. get genome
   - <http://hgdownload.soe.ucsc.edu/goldenPath/mm10/bigZips/chromFa.tar.gz>
   - <ftp://ftp.ensembl.org/pub/release-93/gtf/mus_musculus/Mus_musculus.GRCm38.93.gtf.gz>
1. get raw data
   - <ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR065/SRR065544/SRR065544_1.fastq.gz>
   - <ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR065/SRR065544/SRR065544_2.fastq.gz>
   - <ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR065/SRR065545/SRR065545_1.fastq.gz>
   - <ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR065/SRR065545/SRR065545_2.fastq.gz>
1. make hisat index

   ```bash
   # extract X chromosome sequence
   tar -xz -f chromFa.tar.gz chrX.fa
   mv chrX.fa Mus_musculus_chrX.fa
   
   # use only X chromosome
   zcat Mus_musculus.GRCm38.93.gtf.gz | grep -P '(#!)|(X\t)' > Mus_musculus_chrX.gtf
   
   # make hisat index
   hisat2_extract_splice_sites.py Mus_musculus_chrX.gtf > Mus_musculus_chrX.ss
   hisat2_extract_exons.py        Mus_musculus_chrX.gtf > Mus_musculus_chrX.exon
   
   mkdir hisat2_indexes
   hisat2-2.1.0/hisat2-build -p 4 --ss Mus_musculus_chrX.ss --exon Mus_musculus_chrX.exon Mus_musculus_chrX.fa hisat2_indexes/Mus_musculus_chrX
   ```

1. mapping

   ```bash
   # mapping
   hisat2-2.1.0/hisat2 -p 4 --dta -S SRR065544_chrX.sam -x hisat2_indexes/Mus_musculus_chrX -1 SRR065544_1.fastq.gz -2 SRR065544_2.fastq.gz
   hisat2-2.1.0/hisat2 -p 4 --dta -S SRR065545_chrX.sam -x hisat2_indexes/Mus_musculus_chrX -1 SRR065545_1.fastq.gz -2 SRR065545_2.fastq.gz
   
   # covert to .bam
   samtools sort -@ 4 -o SRR065544_chrX_raw.bam SRR065544_chrX.sam
   samtools sort -@ 4 -o SRR065545_chrX_raw.bam SRR065545_chrX.sam
   
   # filter only mapped reads
   bamtools index -in SRR065544_chrX_raw.bam
   bamtools index -in SRR065545_chrX_raw.bam
   
   bamtools filter -isMapped true -in SRR065544_chrX_raw.bam -out SRR065544_chrX.bam
   bamtools filter -isMapped true -in SRR065545_chrX_raw.bam -out SRR065545_chrX.bam
   ```

## 5) Homework and more

探索 SRR065547 与 SRR065546 之间在 X 染色体上存在 differential splicing 的基因, `.bam` 文件可在 [这里](https://github.com/dongzhuoer/lulab-teaching_book-data/tree/master/AS) 下载



