
# 7.2 motif analysis

## Pipeline

## Data Structure

### getting software & data

HOMER contains a program called `findMotifsGenome.pl` that can find enriched motifs in ChIP-Seq peaks

### input

### output

| file name                    | description                                                                     |
|------------------------------|---------------------------------------------------------------------------------|
| `homerMotifs.all.motifs`     | Simply the concatenated file composed of all the `homerMotifs.motifs<#>` files. |
| `motifFindingParameters.txt` | command used to execute `findMotifsGenome.pl`.                                  |
| `knownResults.txt`           | text file containing statistics about known motif enrichment (open in EXCEL).   |
| `seq.autonorm.tsv`           | autonormalization statistics for lower-order oligo normalization.               |
| `homerResults.html`          | formatted output of de novo motif finding:                                      |


![](../.gitbook/assets/chipseq-motif-output.png)

| file name                 | description                                                                                                                            |
|---------------------------|----------------------------------------------------------------------------------------------------------------------------------------|
| `knownResults.html`       | formatted output of known motif finding.                                                                                               |
| `knownResults/ directory` | contains files for the knownResults.html webpage, including `known<#>.motif` files for use in finding specific instance of each motif. |


## Running Scripts

```bash
findMotifsGenome.pl part.peak sacCer2 part.motif.output -len 8
```

Important parameters


- Region Size  
  (`-size <#>`, `-size <#>,<#>`, `-size given`, default: 200)  
  The size of the region used for motif finding is important.  If analyzing 
      ChIP-Seq peaks from a transcription factor, Chuck would recommend 50 bp 
      for establishing the primary motif bound by a given transcription factor 
      and 200 bp for finding both primary and "co-enriched" motifs for a 
      transcription factor.  When looking at histone marked regions, 500-1000 bp
      is probably a good ideaã€‚

- Motif length  
  (`-len <#>` or `-len <#>,<#>,...`, default 8,10,12)   
  In general, it's best to try out enrichment with shorter lengths 
      (i.e. less than 15) before trying longer lengths.

- Number of motifs to find  
  (`-S <#>`, default 25)  
  Specifies the number of motifs of each length to find. 25 is already quite a bit.

## Tips/Utilities

## Homework and more





